### DISCUSSION TITLE
Executing operation on a single host in parallel

### DISCUSSION THREAD

**Discussion by pioio** [pyinfra-dev/pyinfra#917] [General]
Hi, is there way, or a pattern, for executing some (e.g. long-running) operation in parallel to other operations?

For simplicity, let's consider a single target host.
Let's say I want to download a large file onto a host.
Let's say this will take a longer while.

During this time I would like to continue executing other operations from the same deploy on that same (operations which do not depend on this file being there).

I can imagine one way for achieving this would be to start multiple parallel deploys in parallel. But sometimes that's not the best solution.
Are there other ways around this? Any plans for introducing ability of having a single operation run in the background while other are executing, until the execution reaches some sort of synchronisation primitive (to join the flow)?

Another example where this would be handy would be running large number of API calls from the target host, each API call taking a few seconds, but all of them could in principle be done in parallel in e.g. batches of 10. This would speed up the total execution by a factor of 10x in such a scenario.

**Comment by Fizzadar** [comment]
Currently not possible, but this is a really interesting use-case, certainly a way of doing this would be very handy...

**Comment by v3ss0n** [comment]
quite intresting , should be doable with subprocess.
